### DISCUSSION TITLE
Problem with code generation

### DISCUSSION THREAD

**Discussion by tinusd** [casadi/casadi#3232] [Q&A] [answered]
Hello there!

I am using casadi 3.6.3 for Matlab. I am looking into code generation in order to make my function evaluations faster for my large-scale optimization problem. In a first step, I am following the example of Section 5.1 from the manual to get familiar:

x = MX.sym('x',2);
y = MX.sym('y');
f = Function('f',{x,y},...
{x,sin(y)*x},...
{'x','y'},{'r','q'});
f.generate('gen.c');
type('gen.c')

Up to this point, all is fine: gen.c is being generated with the c code inside.

I then run 
system('gcc -fPIC -shared gen.c -o gen.so') and it outputs:
***********************************************
gen.c:1:0: warning: -fPIC ignored for target (all code is position independent) 
 /* This file was automatically generated by CasADi 3.6.3. 
 ^ 

ans =

     0
***********************************************

When I then run f = external('f', './gen.so'), I get the following error:
***********************************************
Error using casadi.external (line 121)
.../casadi/core/casadi_os.cpp:166: Assertion "handle!=nullptr" failed:
DllLibrary::init_handle: Cannot load shared library './gen.so':
   (
    Searched directories: 1. casadipath from GlobalOptions
                          2. CASADIPATH env var
                          3. PATH env var (Windows)
                          4. LD_LIBRARY_PATH env var (Linux)
                          5. DYLD_LIBRARY_PATH env var (osx)
    A library may be 'not found' even if the file exists:
          * library is not compatible (different compiler/bitness)
          * the dependencies are not found
   )
  Tried 'C:\Users\tinusd\temp\casadi-3.6.3-windows64-matlab2018b' :
    Error code (WIN32): 126
  Tried '' :
    Error code (WIN32): 126
  Tried '.' :
    Error code (WIN32): 193
***********************************************

Any idea what the problem is? Thanks in advance! (And sorry if I am making an obvious mistake...)

PS: when I follow the "jit" example from the nlp_codegen.py, it works. However, when I fllow the "external" example, I get the same error.

**Comment by jgillis** [comment, marked as answer]
The compiler command seems to be for linux, while you are on a windows machine..in your case, you need to produce a gen.dll

**Reply by tinusd** [reply]
Thanks for the quick response jgillis! I see... how embarrassing. I will look into generating a *.dll then. Just to make sure, the first steps---generating gen.c---will stay the same (regardless if I am on Windows)?.

**Reply by jgillis** [reply]
Stays the same yes.

**Reply by tinusd** [reply]
OK, I managed do it using Clang.

system('cl.exe /D_USRDLL /D_WINDLL gen.c /MT /link /DLL /OUT:gen.dll')

Thanks a lot for your tip!
